File: packages/native-helpers/linux-helper-ts/src/rpc/stdout-writer.ts
/**
 * Serialized stdout writer.
 * Both RPC responses and evdev keyboard events write to stdout,
 * so we serialize writes to prevent interleaving.
 * Uses synchronous writes to ensure output is flushed before process exits.
 */

import * as fs from "node:fs";

const STDOUT_FD = 1;

export function writeLine(obj: unknown): void {
  const line = JSON.stringify(obj) + "\n";
  fs.writeSync(STDOUT_FD, line);
}

export function writeResponse(id: string, result: unknown): void {
  writeLine({ id, result });
}

export function writeError(
  id: string,
  code: number,
  message: string,
  data?: unknown,
): void {
  writeLine({ id, error: { code, message, data } });
}

export function writeEvent(event: unknown): void {
  writeLine(event);
}

File: packages/native-helpers/linux-helper-py/linux_helper/stdout_writer.py
"""Thread-safe stdout writer for JSON-RPC responses and events."""

import json
import sys
import threading
from typing import Any

_lock = threading.Lock()


def write_line(obj: Any) -> None:
    line = json.dumps(obj, ensure_ascii=False)
    with _lock:
        sys.stdout.write(line + "\n")
        sys.stdout.flush()


def write_response(request_id: str, result: Any) -> None:
    write_line({"id": request_id, "result": result})


def write_error(request_id: str, code: int, message: str, data: Any = None) -> None:
    error = {"code": code, "message": message}
    if data is not None:
        error["data"] = data
    write_line({"id": request_id, "error": error})


def write_event(event: dict) -> None:
    write_line(event)
